### PR TITLE
refactor: use a common macro to define math UDFs

### LINKED ISSUE #9552: Eliminate code duplication between math UDFs
### Is your feature request related to a problem or challenge?

The codes in [acos.rs](https://github.com/apache/arrow-datafusion/blob/main/datafusion/functions/src/math/acos.rs), [asin.rs ](https://github.com/apache/arrow-datafusion/blob/main/datafusion/functions/src/math/asin.rs) under the new functionn packege and codes in #9535 are quite similar.
Perhaps they could be simplified with the help of a common macro.

### Describe the solution you'd like

A common macro similar to [math_unary_function](https://github.com/apache/arrow-datafusion/blob/88187d4dcc060d4d8ae7c680594e72fdfbc24c6a/datafusion/physical-expr/src/math_expressions.rs#L83), used to define math UDFs.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Closes #9552.

## Rationale for this change
Use a common macro to define unary math UDFs, aiming to reduce code duplication and facilitate subsequent porting.
These functions include
https://github.com/apache/arrow-datafusion/blob/0302d6530a5a7fffe7da40e0f3b13794a9c693ff/datafusion/physical-expr/src/math_expressions.rs#L158-L178

## Are these changes tested?
Yes. By existing tests.

## Are there any user-facing changes?
No

### CODE REVIEW

**Review by alamb — APPROVED** [review]
👨‍🍳  👌 

Very nice @jonahgao

**Comment by alamb** [comment]
This is a really nice design

File: datafusion/functions/src/macros.rs [hunk 1]
```diff
@@ -146,3 +146,102 @@ macro_rules! downcast_arg {
...
+
+                fn signature(&self) -> &Signature {
+                    &self.signature
+                }
+
+                fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+                    let arg_type = &arg_types[0];
+
+                    match arg_type {
+                        DataType::Float64 => Ok(DataType::Float64),
```
Comment (comphead) [inline datafusion/functions/src/macros.rs:205]:
```suggestion
```

**Review by comphead — COMMENTED** [review]
Thanks @jonahgao
nice change

**Comment by jonahgao** [comment]
Thank you for the review! @alamb @comphead

**Review by comphead — APPROVED** [review]
lgtm thanks @jonahgao

**Comment by alamb** [comment]
Merging to get this (good) new pattern estabilished before too many other functions get ported

**Pull request merged**